Feat Followers

SanGeeky opened the pull request:
This PR adds the followers feature, the main changes are:

- Relation many-to-many between `users` and `followers`
- Add Follow and Unfollow users and their respective views
- Show followers and followed users
- Comments only for followers

SanGeeky committed:
Add relation Followers

SanGeeky committed:
Relation many-to-many in Users and Followers

SanGeeky committed:
Feat Follow and Unfollow users

SanGeeky committed:
Show following features in views

- Button to follow user
- Show User followers
- Followed users by users

SanGeeky committed:
Feat comments for followers

comments available only for followers of the creator user

SanGeeky committed:
Rename followers and  user follows

mauriciosabogal commented on app/views/users/show.html.erb:
```diff
@@ -9,6 +9,27 @@
       <%= @user.first_name %> <%= @user.last_name %>
     </span>
+    <br>
+    <% if current_user %>
```
Avoid having too many levels of `if`s. You can refactor this code.

SanGeeky replied:
Done

mauriciosabogal commented on db/migrate/20210428151759_create_followers.rb:
```diff
@@ -0,0 +1,10 @@
+  def change
+    create_table :followers do |t|
+      t.references :user, null: false, foreign_key: true
+      t.bigint :follower
```
Why is `follower` not a foreign key?

SanGeeky replied:
I only added for models now i add the migration for the database.
```
class AddForeignToFollowers < ActiveRecord::Migration[6.1]
  def change
    add_foreign_key :followers, :users, column: :follower_id
  end
end
```

mauriciosabogal commented on app/models/user.rb:
```diff
@@ -4,8 +4,19 @@
   validates :email, presence: true, uniqueness: true
   validates :username, presence: true, uniqueness: true
+
+  def follow_user?(user_id)
```
This method returns a boolean if the user is the same, and an object if is not. The return type should be the same.

SanGeeky replied:
```
 def follow_user?(user_id)
    following = follow_to.find_by(user_id: user_id)
 end
```
I refactor the conditional in view and the method

mauriciosabogal commented on app/controllers/followers_controller.rb:
```diff
@@ -0,0 +1,33 @@
+
+  def create
+    follow = Follower.create(follow_params)
+    follow.save
```
You are saving the object twice.

SanGeeky replied:
```
follow = Follower.new(follow_params)
follow.save if follow.valid?
redirect_article
```
Is true, now I am creating the object and saving

SanGeeky committed:
Refactor code with PR recommendations

- follow creation stored one time
- follow_user? returning nil or object
- refactor conditional in view users/show
- Add foreign key migration in followers model
